src/axom/spin/BVH.hpp

```diff
@@ -0,0 +1,848 @@
+
+  return in_right;
+}
+
+}
+
+//------------------------------------------------------------------------------
+//  BVH IMPLEMENTATION
+//------------------------------------------------------------------------------
+template< int NDIMS, typename ExecSpace, typename FloatType >
+BVH< NDIMS, ExecSpace, FloatType >::BVH( const FloatType* boxes,
+                                         IndexType numItems ) :
+  m_numItems( numItems ),
+  m_boxes( boxes )
+{
+
+}
+
+//------------------------------------------------------------------------------
+template< int NDIMS, typename ExecSpace, typename FloatType >
+BVH< NDIMS, ExecSpace, FloatType >::~BVH()
+{
+  m_bvh.deallocate();
+}
+
+//------------------------------------------------------------------------------
+template< int NDIMS, typename ExecSpace, typename FloatType >
+int BVH< NDIMS, ExecSpace, FloatType >::build()
+{
+  // STEP 0: set the default memory allocator to use for the execution space.
+  umpire::Allocator current_allocator = axom::getDefaultAllocator();
+  const int allocatorID = spin::execution_space< ExecSpace >::allocatorID();
+  axom::setDefaultAllocator( allocatorID );
+
+  // STEP 1: Build a RadixTree consisting of the bounding boxes, sorted
+  // by their corresponding morton code.
+  internal::linear_bvh::RadixTree< FloatType, NDIMS > radix_tree;
+  internal::linear_bvh::AABB< FloatType, NDIMS > global_bounds;
+  internal::linear_bvh::build_radix_tree< ExecSpace >(
+      m_boxes, m_numItems, global_bounds, radix_tree );
+
+  // STEP 2: emit the BVH data-structure from the radix tree
+  m_bvh.m_bounds = global_bounds;
+  m_bvh.allocate( m_numItems );
+
+  // STEP 3: emit the BVH
+  internal::linear_bvh::emit_bvh< ExecSpace >( radix_tree, m_bvh );
+
+  radix_tree.deallocate();
+
+  // STEP 4: restore default allocator
+  axom::setDefaultAllocator( current_allocator );
+  return BVH_BUILD_OK;
+}
+
+//------------------------------------------------------------------------------
+template< int NDIMS, typename ExecSpace, typename FloatType >
+void BVH< NDIMS, ExecSpace, FloatType >::getBounds( FloatType* min,
+                                                    FloatType* max ) const
+{
+  SLIC_ASSERT( min != nullptr );
+  SLIC_ASSERT( max != nullptr );
+  m_bvh.m_bounds.min( min );
+  m_bvh.m_bounds.max( max );
+}
+
+//------------------------------------------------------------------------------
+template< int NDIMS, typename ExecSpace, typename FloatType >
+void BVH< NDIMS, ExecSpace, FloatType >::find( IndexType* offsets,
+                                               IndexType* counts,
+                                               IndexType*& candidates,
+                                               IndexType numPts,
+                                               const FloatType* x,
+                                               const FloatType* y,
+                                               const FloatType* z ) const
+{
+  AXOM_STATIC_ASSERT_MSG( NDIMS==3,
```

Comment:
Is there a reason you chose a `AXOM_STATIC_ASSERT_MSG` instead of template specialization, ie `BVH< 3, ExecSpace, FloatType >::find`

Reply:
Seemed cleaner (easier?) and I can also provide a more meaningful message....